Mathlib/Tactic/GRW/Core.lean

```diff
@@ -0,0 +1,179 @@
+# GRW Tactic
+
+This module defines the core of the `grw` tactic.
+
+-/
+
+open Lean Meta Std.Tactic.LabelAttr Elab Tactic
+
+namespace Mathlib.Tactic.GRW
+
+initialize registerTraceClass `GRW
+
+/--
+Lemmas marked `@[grw]` are used by the `grw` tactic to use a relation to rewrite an expression.
+
+The lemma is used to rewrite its first explicit argument into the result type. The other arguments
+are then filled in using the `gcongr` tactic.
+
+For example, this lemma shows `grw` how to convert `a < b` into `c < d`.
+
+```lean
+@[grw]
+lemma rewrite_lt {α : Type} [Preorder α] {a b c d : α} (h₁ : a < b) (h₂ : c ≤ a) (h₃ : b ≤ d) :
+    c < d := lt_of_le_of_lt h₂ (lt_of_lt_of_le h₁ h₃)
+```
+
+These lemmas can do more than just use transitivity. This lemma shows `grw` to to rewrite `a ∈ X`
+into `a ∈ Y`.
+
+```lean
+@[grw]
+lemma rewrite_mem {α : Type} {a : α} {X Y: Set α} (h₁ : a ∈ X) (h₂ : X ⊆ Y) : a ∈ Y := h₂ h₁
+```
+-/
+initialize ext : LabelExtension ← (
+  let descr := "A lemma describing how to convert the first argument into the target type, possibly
+introducing side goals. These side goals will be solved with `gcongr`"
+  let grw := `grw
+  registerLabelAttr grw descr grw)
+
+open GCongr
+
+/-- Given `relationType := R a b`, returns `(a, b)` -/
+def getNeedleReplacement (relationType : Expr) : MetaM (Expr × Expr) := do
+  let args := relationType.getAppArgs
+  if args.size < 2 then
+    throwError "Expecting relation but got {relationType}"
+  return (args[args.size - 2]!, args[args.size - 1]!)
+
+/--
+Constructs the new expression after rewriting occurrences of the LHS of `rule` with the RHS
+in `oldExpr` (or vice versa when `rev` is true). Returns `(template, newExpr)` where `newExpr` is
+the replaced expression and `template` is the rewrite motive with the bound variable replaced with
+a fresh mvar (used as input to gcongr).
+-/
+def getNewExpr (rule : Expr) (rev : Bool) (oldExpr : Expr) : MetaM (Expr × Expr) := do
+  let ruleType ← instantiateMVars (← inferType rule)
+  let (needle, replacement) ← if rev then
+    Prod.swap <$> getNeedleReplacement ruleType
+  else
+    getNeedleReplacement ruleType
+  trace[GRW] "Got needle = {needle} replacement = {replacement}"
+  let abst ← withReducible <| kabstract oldExpr needle
+  if !abst.hasLooseBVars then
+    throwError "Could not find pattern {needle} in {oldExpr}"
+  let newExpr := abst.instantiate1 replacement
+  trace[GRW] "old expr {oldExpr} new expr {newExpr}"
+  let template := abst.instantiate1 (← mkFreshExprSyntheticOpaqueMVar ruleType)
+  return (← whnfR template, newExpr)
+
+/-- Try to discharge the goal `goal` using `gcongr` using the provided `template` and using only
+`rule` for discharging the main subgoals. -/
+def useRule (template rule : Expr) (goal : MVarId) : MetaM (Array MVarId) := do
+  let template ← instantiateMVars template
+  try
+    if template.hasExprMVar then failure
+    goal.applyRfl
+    trace[GRW] "used reflexivity"
+    return #[]
+  catch _ => pure ()
```

Comment:
This is going to catch all sorts of unwanted things like timeouts, isn't it?

Reply:
It looks like there's no good answer here (though maybe @digama0 knows of one).

Reply:
this is not really something that should be solved here, it's endemic to the whole tactic infrastructure. Maybe we could change the `tryCatch` implementation for `TacticM` to not catch timeouts and other abort-level commands?

Reply:
I think that must be a recent change: thanks for digging!